contributors/devel/sig-architecture/api-conventions.md

```diff
@@ -308,17 +308,67 @@ response reduces the complexity of these clients.
+understand resource-specific status details. Conditions should complement more
+detailed information about the observed status of an object written by a
+controller, rather than replace it. For example, the "Available" condition of a
+Deployment can be determined by examining `readyReplicas`, `replicas`, and
+other properties of the Deployment. However, the "Available" condition allows
+other components to avoid duplicating the availability logic in the Deployment
+controller.
 
 Objects may report multiple conditions, and new types of conditions may be
 added in the future or by 3rd party controllers. Therefore, conditions are
-represented using a list/slice, where all have similar structure.
+represented using a list/slice of objects, where each condition has a similar
+structure. This collection should be treated as a map with a key of `type`.
+
+Conditions are most useful when they follow some consistent conventions:
+
+* Conditions should be added to explicitly convey properties that users and
+  components care about rather than requiring those properties to be inferred
+  from other observations.  Once defined, the meaning of a Condition can not be
+  changed arbitrarily - it becomes part of the API, and has the same backwards-
+  and forwards-compatibility concerns of any other part of the API.
+
+* Controllers should apply their conditions to a resource the first time they
+  visit the resource, even if the `status` is Unknown. This allows other
+  components in the system to know that the condition exists and the controller
+  is making progress on reconciling that resource.
+
+   * Not all controllers will observe the previous advice about reporting
+     "Unknown" or "False" values. For known conditions, the absence of a
+     condition status should be interpreted the same as `Unknown`, and
+     typically indicates that reconciliation has not yet finished (or that the
+     resource state may not yet be observable).
+
+* For some conditions, `True` represents normal operation, and for some
+  conditions, `False` represents normal operation. ("Normal-true" conditions
+  are sometimes said to have "positive polarity", and "normal-false" conditions
+  are said to have "negative polarity".) Without further knowledge of the
+  conditions, it is not possible to compute a generic summary of the conditions
+  on a resource.
+
+* Condition type names should make sense for humans; neither positive nor
+  negative polarity can be recommended as a general rule. A negative condition
+  like "MemoryExhausted" may be easier for humans to understand than
+  "SufficientMemory". Conversely, "Ready" or "Succeeded" may be easier to
+  understand than "Failed", because "Failed=Unknown" or "Failed=False" may
+  cause double-negative confusion.
+
+* Condition type names should describe the current observed state of the
+  resource, rather than describing the current state transitions. This
```

Comment:
Sufficiently slow transitions are best thought of as states in and of themselves, so I'm not convinced about this paragraph.

Reply:
I'm ready to sign off on the other changes, I'm on the fence about whether to ask for a change here or not.

Reply:
I can update to mention slow transitions as states (LetsEncrypt cert provisioning strikes me as an example)

Reply:
Yeah, this paragraph was the only one I bumped on.

I'm thinking of conditions like `Resizing=True` the PV controller or node has contemplated adding in the past, and trying to figure out what the alternative would be if this was the guidance. `NotResizing=True`, etc, to conform to this paragraph seems of the "up with which I will not put" variety.

Reply:
`ResizeInProgress`? That also provides an indication that "yes, I picked up this work, and I'm working on it now".

Reply:
I guess? I don't really see a distinction between `Resizing=True` and `ResizeInProgress=True` and would probably have picked the former.

Reply:
Actually, it seems that `ResizeSucceeded` would be a better condition:

* `True` = object is at the correct size
* `Unknown` = I'm still working
* `False` = I'm sorry, Dave. I'm afraid I can't do that.

Reply:
`Resizing` is fine (and abides by the earlier statement to prefer brevity).

I think `AboutToStartResizing` and `AboutToFinishResizing` are ones we'd want to recommend against.

```
State transitions have transitions
In between to change 'em
Those have yet smaller changes too,
and so, ad infinitum
```

([original](https://en.wikipedia.org/wiki/Siphonaptera_%28poem%29))

Reply:
Added a new paragraph suggesting `Provisoned` or `SizeAllocated` as names for these "takes a while" conditions. Another good example might be certificate signing via the ACME protocol with LetsEncrypt, which can take 5 minutes or so. There's a distinction between "still working" and "it failed due to rate limit" that seems like it could be usefully captured within a single condition.